cpp/ql/src/experimental/Security/CWE/CWE-758/DangerousUseMallocWithZeroSize.ql

```diff
@@ -0,0 +1,164 @@
+/**
+ * @name Dangerous use malloc with zero size.
+ * @description The behavior of the malloc function is not defined when the size value is zero.
+ * @kind problem
+ * @id cpp/dangerous-use-malloc-with-zero-size
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-758
+ */
+
+import cpp
+import semmle.code.cpp.commons.Exclusions
+import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+
+/** Holds if there are checks above the call */
+predicate existsChecks(FunctionCall fc) {
+  (
+    exists(AssignBitwiseOperation ab |
+      ab.getLValue().(VariableAccess).getTarget() = fc.getArgument(0).(VariableAccess).getTarget() and
+      ab.getASuccessor*() = fc
+    )
+    or
+    exists(Initializer it |
+      fc.getArgument(0).(VariableAccess).getTarget().getInitializer() = it and
+      (
+        it.getExpr().getAChild*() instanceof AddExpr or
+        it.getExpr().isConstant()
+      )
+    )
+    or
+    exists(Assignment ae |
+      ae.getLValue().(VariableAccess).getTarget() = fc.getArgument(0).(VariableAccess).getTarget() and
+      (
+        ae.getRValue().getAChild*() instanceof AddExpr or
+        ae.getRValue().isConstant()
+      ) and
+      ae.getASuccessor*() = fc
+    )
+    or
+    exists(IfStmt ifs, ComparisonOperation co, Expr ec |
+      ifs.getCondition().getAChild*() = co and
+      co.hasOperands(fc.getArgument(0).(VariableAccess).getTarget().getAnAccess(), ec) and
+      exists(Expr etmp |
+        (
+          etmp.getValue() = "0" or
+          etmp.getValue() = "1"
+        ) and
+        (
+          ec = etmp or
+          globalValueNumber(ec) = globalValueNumber(etmp)
+        )
+      )
+    )
+  )
+}
+
+/** Holds if the function within which the call is found is not called elsewhere. */
+predicate enclosingFunctionNotCall(FunctionCall fc) {
+  exists(Function fn |
+    fn = fc.getEnclosingFunction() and
+    not exists(fn.getACallToThisFunction()) and
+    fn.getAParameter().getAnAccess() = fc.getArgument(0) and
+    not exists(TemplateClass tc | tc.getACanonicalMemberFunction() = fn)
+  )
+}
+
+/** Holds if buffer and length are used in string functions or arrays. */
+predicate dangerousUseBufferAndSize(FunctionCall fc) {
+  exists(Expr e0 |
+    fc.getASuccessor*() = e0 and
+    (
+      e0 instanceof ArrayExpr and
```

Comment:
I find it very difficult to read this code without any comments. It looks like there are two cases in here: One for `ArrayExpr`s and one for a bunch of string-related functions. Can you add a comment saying what each case is looking for?

Reply:
i fixed it. Have a look, please

Reply:
Thanks!